fix(provider/kubernetes): exclude stages when k8s v2 is the only provider

sbwsg opened the pull request:
~Providers are passed around as strings when assessing whether stages support them or not.  Both the kubernetes v1 and the kubernetes v2 provider share a provider string of `kubernetes`.  Therefore the stage filtering assumes that any stage supported by v1 is also supported by v2 but this is not the case.~

~This workaround allows a field to be added to a stage that lists specific provider versions that it doesn't support.  After the initial set of stages is filtered via provider strings, a second round of filtering is performed to remove any stage where the set of accounts for this application are entirely excluded by the stage's type.~

~This solution doesn't feel quite right but I can't see a way past this problem without a large refactor such that everywhere we pass a provider string we also pass around a version.  Maybe that would be a better solution though.  Open to ideas here.~

Fixed.

#### Before: (A Deploy stage appears in an app with only a k8s v2 account configured)

![before](https://user-images.githubusercontent.com/34253460/36002446-376b2378-0cf8-11e8-8218-c58725aca041.png)

#### After: (The Deploy stage is filtered out because its excludedCloudProviders field includes k8s v2)

![after](https://user-images.githubusercontent.com/34253460/36002466-4686c5f6-0cf8-11e8-9418-a77d1c4477c8.png)

https://github.com/spinnaker/spinnaker/issues/2277

lwander commented:
Nice!

danielpeach commented on app/scripts/modules/core/src/pipeline/config/stages/stage.module.js:
```diff
@@ -52,8 +53,12 @@ module.exports = angular.module('spinnaker.core.pipeline.config.stage', [
+    accountService.providerAccounts($scope.application).then(providerAccounts => {
+      const providers = Object.keys(providerAccounts);
+      const accounts = providers.reduce((memo, provider) => memo.concat(providerAccounts[provider]), []);
+      let stageTypes = pipelineConfig.getConfigurableStageTypes(providers);
```
My slight preference would be that `getConfigurableStageTypes` returned the exact set of `stageTypes` to render (i.e., this logic should be encapsulated in that method, probably by passing it a list `{ provider: string, providerVersion: string}[]`).

sbwsg replied:
:+1: done, thanks, def the better approach.

danielpeach commented on app/scripts/modules/core/src/pipeline/config/stages/stage.module.js:
```diff
@@ -197,6 +202,20 @@ module.exports = angular.module('spinnaker.core.pipeline.config.stage', [
 
+    function stageExcludesAllAccounts(stageType, accounts) {
+      if (stageType.excludedCloudProviders) {
+        const anyNotExcluded = some(accounts, account => {
```
Would use native `[].some` here and below.

danielpeach replied:
Would also be a little more readable if the bool were swapped such that this was an `[].every`. i.e., every `account` is excluded by at least one of the `excludedCloudProviders`.

sbwsg replied:
done

danielpeach commented on app/scripts/modules/core/src/pipeline/config/stages/stage.module.js:
```diff
@@ -197,6 +202,20 @@ module.exports = angular.module('spinnaker.core.pipeline.config.stage', [
+    function stageExcludesAllAccounts(stageType, accounts) {
+      if (stageType.excludedCloudProviders) {
+        const anyNotExcluded = some(accounts, account => {
+          return some(stageType.excludedCloudProviders, excluded => (
```
Could make this `stageType.excludedCloudProviders || []` and remove the entire enclosing `if { ... } else { ... }`.

sbwsg replied:
done

danielpeach commented on app/scripts/modules/core/src/account/account.service.ts:
```diff
@@ -158,6 +162,16 @@ export class AccountService {
   }
 
+  public providerAccounts(application: Application = null): Promise<IProviderAccounts> {
+    const result: Promise<IProviderAccounts> = Promise.all([this.listProviders(application), this.listAccounts()]).then(([providers, accounts]) => {
```
Would return directly on this line.

sbwsg commented:
@danielpeach I've fixed the issues you highlighted with multiple provider versions and removed the custom interfaces in favor of an array of `IAccountDetails`.
